src/dfx/src/lib/nns/install_nns.rs

```diff
@@ -1,32 +1,661 @@
+        install_canister(env, agent, canister_name, &local_wasm_path).await?;
+    }
+    // ... and configure the backend NNS canisters:
+    set_xdr_rate(1234567, &nns_url)?;
+    set_cmc_authorized_subnets(&nns_url, &subnet_id.to_string())?;
+
+    Ok(())
+}
+
+/// Gets a URL, trying repeatedly until it is available.
+pub async fn get_with_retries(url: &Url) -> anyhow::Result<reqwest::Response> {
+    /// The time between the first try and the second.
+    const RETRY_PAUSE: Duration = Duration::from_millis(200);
+    /// Intervals will increase exponentially until they reach this.
+    const MAX_RETRY_PAUSE: Duration = Duration::from_secs(5);
+
+    let mut waiter = Delay::builder()
+        .exponential_backoff_capped(RETRY_PAUSE, 1.4, MAX_RETRY_PAUSE)
+        .build();
+
+    loop {
+        match reqwest::get(url.clone()).await {
+            Ok(response) => {
+                return Ok(response);
+            }
+            Err(err) => waiter.wait().map_err(|_| err)?,
+        }
+    }
+}
+
+/// Gets the local replica type from dfx.json
+///
+/// # Errors
+/// Returns an error if the replica type could not be determined.  Possible reasons include:
+/// - There is no `dfx.json`
+/// - `dfx.json` could not be read.
+/// - `dfx.json` is not valid JSON.
+/// - The replica type is not defined for the `local` network.
+///
+/// # Panics
+/// This code is not expected to panic.
+#[context("Failed to determine the local replica type.")]
+fn local_replica_type(env: &dyn Environment) -> anyhow::Result<ReplicaSubnetType> {
+    Ok(env
+        .get_network_descriptor()
+        .local_server_descriptor()?
+        .replica
+        .subnet_type
+        .unwrap_or_default())
+}
+
+/// Checks that the local replica type is 'system'.
+///
+/// Note: At present dfx runs a single local replica and the replica type is taken from dfx.json.  It is unfortunate that the subnet type is forced
+/// on the other canisters, however in practice this is unlikely to be a huge problem in the short term.
+///
+/// # Errors
+/// - Returns an error if the local replica type in `dfx.json` is not "system".
+/// # Panics
+/// This code is not expected to panic.
+#[context("Verifying that the local replica type is 'system'.")]
+pub fn verify_local_replica_type_is_system(env: &dyn Environment) -> anyhow::Result<()> {
+    match local_replica_type(env) {
+        Ok(ReplicaSubnetType::System) => Ok(()),
+        other => Err(anyhow!("The replica subnet_type needs to be \"system\" to run NNS canisters.  Current value: {other:?}")),
+    }
+}
+
+/// Downloads a file
+#[context("Failed to download {:?} to {:?}.", source, target)]
+pub async fn download(source: &Url, target: &Path) -> anyhow::Result<()> {
+    let buffer = reqwest::get(source.clone()).await?.bytes().await?;
+    let tmp_dir = tempfile::Builder::new().tempdir()?;
+    let downloaded_filename = {
+        let filename = tmp_dir.path().join("wasm");
+        let mut file = fs::File::create(&filename)?;
```

Comment:
We've run into a lot of problems with rust errors not showing enough context to be able to tell which part failed.
```suggestion
        let mut file = fs::File::create(&filename).with_context(||format!("Failed to create file {}', filename.display()))?;
```